### PR TITLE
0.2.24

### PR DESCRIPTION
**Pull request by zhfnjust**
1. add auto gen debug launch.json  when failed , Fix sCrypt-Inc/ide#142
2. Locate sCrypt source where scrypttest fails Fix sCrypt-Inc/ide#140

### CODE REVIEW

File: src/contract.ts [hunk 1]
```diff
@@ -32,6 +36,8 @@ export class AbstractContract {
   public static abi: ABIEntity[];
   public static asm: string;
   public static abiCoder: ABICoder;
+  public static debugAsm?: DebugModeAsmWord[];
```
Comment (xhliu) [inline src/contract.ts:39]: Rename debugAsm -> opcodes. There is no debug here. And it's an array.

Let's also change the original DebugModeAsmWord -> OpCode

File: src/contract.ts [hunk 2]
```diff
@@ -58,21 +64,67 @@ export class AbstractContract {
...
+    
+    const debugAsm: DebugModeAsmWord[] =  Object.getPrototypeOf(this).constructor.debugAsm;
+
+    const result = bsi.verify(us, ls, tx, inputIndex, DEFAULT_FLAGS, new bsv.crypto.BN(inputSatoshis));
+
+    let error = bsi.errstr;
+
+    const offset = unlockingScriptASM.trim().split(' ').length;
+    
+    // the complete script will have op_return and data, but debugAsm do not have, so we need to make sure the index in bound.
```
Comment (xhliu) [inline src/contract.ts:101]: // the complete script may have op_return and data, but compiled output does not have it. So we need to make sure the index is in boundary.

File: src/contract.ts [hunk 3]
```diff
@@ -58,21 +64,67 @@ export class AbstractContract {
...
+    const debugAsm_pc = Math.min(stepCouter -  offset,  debugAsm.length -1);
+
+    if(!result && debugAsm && debugAsm[debugAsm_pc]) {
+
+      let asmWord = debugAsm[debugAsm_pc]; 
+
+      if(!asmWord.file || asmWord.file === "std") {
+
+        let asmWordMappable  = AbstractContract.findMappableAsmWord(debugAsm, debugAsm_pc);
```
Comment (xhliu) [inline src/contract.ts:110]: findMappableAsmWord -> findSrcInfo

**Review by xhliu — CHANGES_REQUESTED** [review]
💯 

Code works. Let's refactor it.

File: src/contract.ts (hunk 3 above)
Comment (zhfnjust) [reply src/contract.ts:110]: fixed

File: src/contract.ts (hunk 2 above)
Comment (zhfnjust) [reply src/contract.ts:101]: fixed

File: src/contract.ts (hunk 1 above)
Comment (zhfnjust) [reply src/contract.ts:39]: fixed

File: src/contract.ts [hunk 4]
```diff
@@ -58,21 +64,67 @@ export class AbstractContract {
...
+    const debugAsm: DebugModeAsmWord[] =  Object.getPrototypeOf(this).constructor.debugAsm;
+
+    const result = bsi.verify(us, ls, tx, inputIndex, DEFAULT_FLAGS, new bsv.crypto.BN(inputSatoshis));
+
+    let error = bsi.errstr;
+
+    const offset = unlockingScriptASM.trim().split(' ').length;
+    
+    // the complete script will have op_return and data, but debugAsm do not have, so we need to make sure the index in bound.
+    const debugAsm_pc = Math.min(stepCouter -  offset,  debugAsm.length -1);
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply src/contract.ts:102]: fixed

File: src/utils.ts [hunk 5]
```diff
@@ -467,3 +470,91 @@ export function checkStruct(s: StructEntity, arg: Struct): void {
...
+  const result = compile(
+    { path: file },
+    { desc: true, debug: true, outputDir: out ? out : join(__dirname, '../out'),
+		  cmdPrefix: scryptc
+    }
+  );
+
+  if (result.errors.length > 0) {
+    console.log(`Contract ${file} compiling failed with errors:`);
+    console.log(result.errors);
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply src/utils.ts:555]: fix

File: src/utils.ts [hunk 6]
```diff
@@ -467,3 +470,91 @@ export function checkStruct(s: StructEntity, arg: Struct): void {
...
+    case 'OP_NUMEQUALVERIFY':
+    case 'OP_NIP':
+      return true;
+    default:
+      return false;
+  }
+}
+
+
+function findCompiler(directory) {
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply src/utils.ts:508]: Fixed

File: src/utils.ts [hunk 7]
```diff
@@ -467,3 +470,91 @@ export function checkStruct(s: StructEntity, arg: Struct): void {
...
+      return false;
+    }
+    return true;
+  });
+
+  return result;
+}
+
+
+export function isEmpty(obj: unknown) {
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply src/utils.ts:489]: Fixed

File: src/abi.ts [hunk 8]
```diff
@@ -91,9 +111,73 @@ export class FunctionCall {
...
+    const constructorParams:SupportedParamType[] = this.contract.scriptedConstructor.params;
+
+      const entryMethodParams:SupportedParamType[] = this.params;
+      const entryMethod: string = this.methodName;
+      const name =  `Debug ${Object.getPrototypeOf(this.contract).constructor.contractName}`;
+      const program = `${Object.getPrototypeOf(this.contract).constructor.file}`;
+
+      let debugConfig: DebugConfiguration = {
+        type: "scrypt",
+        request: "launch",
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply src/abi.ts:128]: debug config need this, and they fixed

File: test/contract.test.ts [hunk 9]
```diff
@@ -12,11 +12,11 @@ const inputSatoshis = 100000;
 const tx = newTx(inputSatoshis);
 const txContext = { inputSatoshis, tx };
 
-const jsonDescr = loadDescription('p2pkh.scrypt');
+const jsonDescr = loadDescription('p2pkh_desc.json');
 
 describe('buildContractClass()', () => {
 
-  const DemoP2PKH = buildContractClass(jsonDescr);
+  const DemoP2PKH = buildContractClass(compileContract(loadFile('p2pkh.scrypt')));
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply test/contract.test.ts]: see  63b093f9004e6e5d122fb2aa1727b267d73707da, change to use jsonDescr

File: src/abi.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (freedomhero) [reply src/abi.ts:128]: could change DebugConfiguration interface to :
```
export interface DebugConfiguration {
  type: 'scrypt';
  name: string;
  request: 'launch';
  program: string;
  constructorParams: SupportedParamType[];
  entryMethod: string;
  entryMethodParams: SupportedParamType[];
  txContext?: any;
}
```

File: src/abi.ts [hunk 10]
```diff
@@ -91,9 +111,73 @@ export class FunctionCall {
     return this.toScript().toHex();
   }
 
+  
+
+  genLaunch(txContext?: TxContext): string{
```
Comment (freedomhero) [inline src/abi.ts:116]: type FileUri = string;
genLaunchConfigFile(...) : FileUri

File: src/abi.ts [hunk 11]
```diff
@@ -91,9 +111,73 @@ export class FunctionCall {
...
+      if(this.contract.dataPart) {
+        Object.assign(debugTxContext, {opReturn: this.contract.dataPart.toASM()});
+      }
+
+      if(!isEmpty(debugTxContext)) {
+        Object.assign(debugConfig, {txContext: debugTxContext});
+      }
+
+      let launch: DebugLaunch = {
+        version: "0.2.0",
```
Comment (freedomhero) [inline src/abi.ts:158]: version also could be fixed to '0.2.0' in interface declaration.

File: src/utils.ts [hunk 12]
```diff
@@ -467,3 +469,74 @@ export function checkStruct(s: StructEntity, arg: Struct): void {
...
+}
+
+
+
+function getCIScryptc(): string | undefined {
+   const scryptc =  findCompiler(__dirname);
+  return fs.existsSync(scryptc) ? scryptc : undefined;
+}
+
+export function compileContract(file: string, out?: string) {
```
Comment (freedomhero) [inline src/utils.ts:515]: Missing return type declaration

Comment (zhfnjust) [reply src/utils.ts:515]: Fixed

File: src/abi.ts (hunk 11 above)
Comment (zhfnjust) [reply src/abi.ts:160]: Fixed

File: src/abi.ts (hunk 10 above)
Comment (zhfnjust) [reply src/abi.ts]: Fixed

File: test/helper.ts [hunk 13]
```diff
@@ -4,13 +4,17 @@ import { bsv } from '../src/utils';
...
 export function loadDescription(fileName: string): ContractDescription {
-  return JSON.parse(readFileSync(join(__dirname, 'fixture', fileName.replace('.scrypt', '_desc.json'))).toString());
+  return JSON.parse(readFileSync(join(__dirname, "../out/", fileName)).toString());
 }
 
 export function loadASM(fileName: string): string {
   return readFileSync(join(__dirname, 'fixture', fileName.replace('.scrypt', '_asm.json'))).toString();
 }
 
+export function loadFile(fileName: string): string {
```
Comment (freedomhero) [inline test/helper.ts:14]: rename to getContractFilePath

File: src/contract.ts [hunk 14]
```diff
@@ -58,21 +66,63 @@ export class AbstractContract {
...
+  
+      if(!result && opcodes[pc]) {
+  
+        let opcode = opcodes[pc]; 
+  
+        if(!opcode.file || opcode.file === "std") {
+  
+          let srcInfo  = AbstractContract.findSrcInfo(opcodes, pc);
+  
+          opcode.file = srcInfo.file;
```
Comment (freedomhero) [inline src/contract.ts:115]: opcode.file = srcInfo?.file

File: src/contract.ts [hunk 15]
```diff
@@ -58,21 +66,63 @@ export class AbstractContract {
...
+          opcode.line = srcInfo.line;
```
Comment (freedomhero) [inline src/contract.ts:116]: opcode.file = srcInfo?.file;
opcode.line = srcInfo?.line;

File: src/scryptTypes.ts [hunk 16]
```diff
@@ -404,6 +408,19 @@ export class Struct extends ScryptType {
     return `Struct(${l})`;
   }
 
+  toJSON() {
```
Comment (freedomhero) [inline src/scryptTypes.ts:411]: Rename to `toJSONObject`

File: src/scryptTypes.ts [hunk 17]
```diff
@@ -38,6 +38,10 @@ export abstract class ScryptType {
     return this._asm;
   }
 
+  toJSON(): string | unknown {
```
Comment (freedomhero) [inline src/scryptTypes.ts:41]: Usually `toJSON` method returns serialized JSON string like "{\"a\":1}" or a JSON object.

File: src/contract.ts (hunk 15 above)
Comment (zhfnjust) [reply src/contract.ts]: Fixed  f707a06dcb0373c0d14f77562f9eaddeca1e60c1

File: test/helper.ts (hunk 13 above)
Comment (zhfnjust) [reply test/helper.ts]: bbb5e69d6e4755c8e0ad447e245a133a66923992

File: src/utils.ts (hunk 12 above)
Comment (zhfnjust) [reply src/utils.ts]: 2333f9602926a2cf7ac2f54283a95702a04d0174

**Pull request merged**